### ISSUE TITLE
db: TestSharedObjectDeletePacing is flaky

### ISSUE THREAD

**Issue by nicktrav** [cockroachdb/pebble#2630]
Seen here: https://github.com/cockroachdb/pebble/actions/runs/5247409884/jobs/9477540951

Reproduces on Linux, locally:

```console
$ go test -run TestSharedObjectDeletePacing -count 1 -exec 'stress ' -timeout 0 -test.v .
...
=== RUN   TestSharedObjectDeletePacing
2023/06/12 16:20:38 shared storage configured; no creatorID yet
2023/06/12 16:20:38 shared storage creatorID set to 1
    compaction_test.go:3979: compaction timed out, possibly due to incorrect deletion pacing
--- FAIL: TestSharedObjectDeletePacing (60.00s)
FAIL


ERROR: exit status 1

1m10s: 8200 runs so far, 14 failures (0.17%)
```

**Comment by nicktrav** [comment]
cc: @RaduBerinde

**Comment by RaduBerinde** [comment]
@jbowens In this test I call `Set + Compact`  for a bunch of keys, then I `DeleteRange + Compact` a subset of those keys, expecting some sstables to be deleted. Sometimes the compaction doesn't result in any obsolete tables. Do I have the wrong expectations about what `Compact()` is supposed to do? Can you think off the top of your head what am I missing?

https://github.com/cockroachdb/pebble/blob/1e8aa51077f0b07d52df681acc790a9020bf0bc9/compaction_test.go#L3955

**Comment by jbowens** [comment]
The test looks reasonable to me.

The [comparison of the number of objects](https://github.com/cockroachdb/pebble/blob/1e8aa51077f0b07d52df681acc790a9020bf0bc9/compaction_test.go#L3967-L3972) broadly has the potential to be brittle, because the number of sstables and where they're split is very timing dependent. Range tombstones can be fragmented into several sstables and may linger in the LSM if they're move-compacted. But no particular scenario comes to mind that would actually preserve `numKeys-2` sstables.

An alternative structure could add a `TableDeleted` event listener to ensure that tables are being deleted (and maybe verify in the provider that they're gone):
```
		EventListener: &EventListener{
			TableDeleted: func(info TableDeleteInfo) {
				// ...
			},
		},
```

**Comment by RaduBerinde** [comment]
Yes, I can make it less brittle by comparing the before and after filenums. Unfortunately, every now and then the compaction doesn't delete any tables (so it's not a matter of having a brittle check); I'll try to figure out why.

**Comment by RaduBerinde** [comment]
It looks like the compaction does delete the files from the new version but the files themselves don't end up being deleted. Something must be holding on to the previous version perhaps..

**Comment by RaduBerinde** [comment]
It looks like `collectTableStats` is running at the same time with the compaction. That causes the `readState`  to have an outstanding reference when it is unrefed by the compaction. When the table stats collection finishes, we do `readState.unref` and that runs `maybeScheduleObsoleteTableDeletion`. But in the latter `acquireCleaningTurn` fails because the compaction is doing `deleteObsoleteFiles` (even though it has nothing to delete).

It seems that we have a bug here.. Any suggestions on how to fix? In `maybeScheduleObsoleteTableDeletion`, we could move `acquireCleaningTurn` into the goroutine and set `waitForOngoing=true`?

**Comment by jbowens** [comment]
My mind melts every time I try to wrap my head around `maybeScheduleObsoleteTableDeletion`, `deleteObsoleteFiles`/`doDeleteObsoleteFiles`, `acquireCleaningTurn`/`releaseCleaningTurn`, `enableFileDeletions`/`disableFileDeletions`, `paceAndDeleteObsoleteFiles` and how they all interact.

Not that we need to do it now, but maybe we should consider a dedicated cleaning goroutine.

**Comment by RaduBerinde** [comment]
I agree - I'll give it a shot.
